### ISSUE TITLE
Test code oddity?

### ISSUE THREAD

**Issue by ajs6f** [trellis-ldp/trellis#219]
I'm implementing the `LdpBasicContainerTests` for `trellis-cassandra` -- only one that relates directly to containment is still broken: `testDeleteContainer`. After a lot of logging, I noticed that the IRIs being used for container and child look odd. Then I looked at the tests that are passing, e.g. `testCreateContainerViaPost` and noticed something different about them. At the [corresponding assertion](https://github.com/trellis-ldp/trellis/blob/master/components/test/src/main/java/org/trellisldp/test/LdpBasicContainerTests.java#L260) the IRI of the child (that we want to see in the object slot of an `ldp:contains` triple) is a reference that [we set above](https://github.com/trellis-ldp/trellis/blob/master/components/test/src/main/java/org/trellisldp/test/LdpBasicContainerTests.java#L249) whereas in the still-failing test we use the `getChildLocation()` method [directly](https://github.com/trellis-ldp/trellis/blob/master/components/test/src/main/java/org/trellisldp/test/LdpBasicContainerTests.java#L348)…

Is that right? Is there a reason that we're not taking a reference at the top of the test? I could be totally nuts here, but it's really strange to me to _that_ assertion failing in _that_ test, because no deletion has yet occurred at that point and the more-or-less the same test code seems to be passing nicely in the other tests.

**Comment by acoburn** [comment]
Is [this](https://github.com/trellis-ldp/trellis/blob/master/components/test/src/main/java/org/trellisldp/test/LdpBasicContainerTests.java#L348) where the test is failing?

**Comment by ajs6f** [comment]
[Exactimundo](https://www.youtube.com/watch?v=_sqrmWFAKC0).

**Comment by ajs6f** [comment]
I can try to clean up my debugging logging (which shows the use of `setChildLocation`, `getChildLocation`, etc. and gist it, if needed.

**Comment by ajs6f** [comment]
Okay, please find that log [here](https://gist.github.com/ajs6f/7285715941d1e7dbd122c9e8d17024c1). It's a long 'un, but it's `TRACE`-level logging for just the one `testDeleteContainer` . Look at the bottom for the blowout. All throughout you can see lines like `DEBUG 14:24:57 Container location is: ...` or `DEBUG 14:24:56 Container location set to: ...` which are indeed logging lines from `setContainerLocation`, `getContainerLocation`, etc.

**Comment by ajs6f** [comment]
I will now try altering the `testDeleteContainer` to look like the others and see what that does.

**Comment by ajs6f** [comment]
I give up. I have the following code:
```
    /**
     * Test deleting a basic container.
     * 
     * @throws Exception
     */
    @Test
    @DisplayName("Test deleting a basic container")
    default void testDeleteContainer() throws Exception {
        final RDF rdf = getInstance();
snipped
```
and Gradle is throwing:
```
> Task :trellis-test:checkstyleMain
[ant:checkstyle] [ERROR] /Users/ajs6f/Documents/trellis/trellis/components/test/src/main/java/org/trellisldp/test/LdpBasicContainerTests.java:341:47: Expected @throws tag for 'Exception'. [JavadocMethod]

```
I don't know if this Gradle, but it looks like Gradle is running checkstyle by using _Ant_ (?! how the *&*(^&(*^ is Ant involved with any of this?).

**Comment by ajs6f** [comment]
Okay, newbie foolishness on my part. A comment works:
```
    /**
     * Test deleting a basic container.
     * 
     * @throws Exception a freaking exception, why is checkstyle even running these rules in test code
     * 
     */
```
Between 0.8 and 1.0, I feel _very strongly_ that we should relax a considerable portion of the checkstyle rules, especially in test code. Requiring descriptions of `Exception` thrown out of tests doesn't seem to me to improve the codebase at all, but actually to make it harder to read.

**Comment by ajs6f** [comment]
I now think there is definitely some kind of test-to-test interaction fouling up here. When I run all of the `LdpBasicContainerTests` together in my impl class, I get three failures:

```
[ERROR] Failures:
[ERROR]   BasicContainmentIT.testCreateContainerViaPost Check that the ETag has been changed ==> expected: not equal but was: <W/"96293535970b4a36beab68d902d29ee2">
[ERROR]   BasicContainmentIT.testCreateContainerViaPut Check ETags 3 and 4 ==> expected: not equal but was: <W/"96293535970b4a36beab68d902d29ee2">
[ERROR]   BasicContainmentIT.testDeleteContainer Check for the presence of an ldp:contains triple ==> expected: <true> but was: <false>
[INFO]
[ERROR] Tests run: 8, Failures: 3, Errors: 0, Skipped: 1
```
I don't know what's being skipped-- I'm not `@Disable`ing anything. When I override all but one, to knock out all but (e.g.) `testCreateContainerViaPost`, it passes:
```
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.657 s - in edu.si.trellis.cassandra.BasicContainmentIT
```
I don't know what to make of this. Have you ever seen this kind of crosstalk?

**Comment by ajs6f** [comment]
Just to be complete and clear, `testCreateContainerViaPost` and `testCreateContainerViaPut` pass on their own or with `testDeleteContainer`, but when run with each other, `testCreateContainerViaPost` fails and `testCreateContainerViaPut` passes, and `testDeleteContainer` does not pass under any combination that I can devise. 😖

**Comment by acoburn** [comment]
I still need to look closely at the gist you provided, but there is definitely some cross-test interactions -- this is why the `@Before*` annotations are used to set up an environment. And yes, checkstyle would complain because this "test" code is actually "main" code for the testing module. But we can certainly relax those rules for this module.

**Comment by ajs6f** [comment]
Okay, let me know what more I can provide logging- or testing-wise.

From my POV, cross-test interactions are okay (not ideal) in internal test code, but if we're going to offer `trellis-test` as the normal bar to meet for a backend (and I would love to, since the LDP test suite is quite stale and getting no fresher) I think we may want to refactor to get slower, completely independent tests. Ideally, all the tests could run in parallel, and I don't think that's an impossible ideal. We just need to have each work in a fresh portion of repo, and do as much as needed of its own setup.
Otherwise, it's very difficult to debug, as I am discovering.

**Comment by acoburn** [comment]
Yes, I completely agree. There are a lot of ways to improve that test code, and anything we can do to make those tests more robust and easier to understand is going to be the right direction.

**Comment by ajs6f** [comment]
Okay, I've got JSON-LD WG right now, so I'll wait to hear from you for next moves.

**Comment by acoburn** [comment]
I think the best solution here would be one of simplicity -- move all of the creation-deletion logic into the test case itself. That is, rather than creating a test resource in the setup phase and then checking that it is deleted in the test phase, do the `POST` inside the test, check that the resource was created, then `DELETE` it, check that it's gone (either `401` or `410`) and, finally, check that the parent no longer `ldp:contains` that resource.

I'll add a commit to that effect as time allows (either this evening or early next week)

**Comment by ajs6f** [comment]
I know that's more verbose, but I agree that it's stronger, at least for now. The other ingredient would seem to be selecting non-overlapping areas of repo in which each test can run, but UUIDs ought to be good enough for that.

**Comment by ajs6f** [comment]
It might be a good idea to add a property to every container that a test starts with (its section of the repo) valued with the `className.methodName` of the test. Then it becomes a bit easier to pause testing to peek in the repo and inspect the contents.

**Comment by ajs6f** [comment]
Actually, I suppose the containers themselves could just be created as `/className.methodName`, because those can't overlap. Then you end up with a nice catalog of test output right there at the root of the repo.

**Comment by acoburn** [comment]
Using a property like that makes a lot of sense. We'd probably still need a unique container name at the root level just so it isn't _required_ that the implementation clean up all the resources between tests.

**Comment by ajs6f** [comment]
#229 looks good, and it has the correct idea in it to close this ticket. Do you want to close now or after we segregate the rest of the tests?

**Comment by ajs6f** [comment]
Hang on-- perhaps I spoke too soon. Are ETags handled the same way in the tests? (From outside each test?) Looks like it, and it looks like I might be having the same problem with them. I'll rewrite a test with this in mind to see if that can pass it.
